File: 5-Problem-A-Day/83. Remove Duplicates from Sorted List.java
/**
 * Definition for singly-linked list. public class ListNode { int val; ListNode next; ListNode(int
 * x) { val = x; } }
 */
class Solution {
    public ListNode deleteDuplicates(ListNode head) {
        if (head == null)
            return null;
        ListNode ans = head, prev = head;
        int count = 1, curr = head.val;
        head = head.next;
        while (head != null) {
            if (head.val != curr) {
                curr = head.val;
                prev.next = head;
                prev = head;
            }
            head = head.next;
        }
        prev.next = null;
        return ans;
    }
}

File: 2021/83. Remove Duplicates from Sorted List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* deleteDuplicates(ListNode* head) {
        if (head == NULL || head->next == NULL) return head;
        ListNode *prev = head, *curr = head->next;        
        while (curr != NULL) {            
            if (prev->val == curr->val) {
                ListNode* toDelete = curr;
                curr = curr->next;
                delete toDelete;
            } else {
                prev->next = curr;
                prev = curr;   
                curr = curr->next;
            }
        }
        prev->next = NULL;
        return head;
    }
};

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* deleteDuplicates(ListNode* head) {
        int curr = 101;
        ListNode* prev = head, *oldHead = head;
        while (head != NULL) {
            if (head->val == curr) {
                prev->next = head->next;   
                delete head;
                head = prev->next;                
            } else {
                curr = head->val;                
                prev = head;
                head = head->next;
            }            
        }
        return oldHead;
    }
};